compiler/rustc_codegen_llvm/src/back/lto.rs

```diff
@@ -325,6 +325,20 @@ fn fat_lto(
         drop(linker);
         save_temp_bitcode(&cgcx, &module, "lto.input");
 
+        // Fat LTO also suffers from the invalid DWARF issue similar to Thin LTO.
+        // Here we rewrite all `DICompileUnit` pointers if there is only one `DICompileUnit`.
+        // This only works around the problem when codegen-units = 1.
+        // Refer to the comments in the `optimize_thin_module` function for more details.
+        let mut cu1 = ptr::null_mut();
+        let mut cu2 = ptr::null_mut();
+        unsafe { llvm::LLVMRustLTOGetDICompileUnit(llmod, &mut cu1, &mut cu2) };
+        if !cu2.is_null() {
```

Comment:
The Thin LTO use of this seems to have the opposite check? It only proceeds if cu2 is nul.

Reply:
Oops... I didn't mean to do the opposite, actually.

However, after the patch, rustc seems to generate valid dwarf for binaries (such as `ripgrep`) while using full LTO. So I truly misunderstood something.

I am not sure how much missing debug info it will increase. But at least Linux perf does not show unknown userspace addresses in my practice...

Reply:
I tried it out. It seems to be doing what I expect (puts everything in one huge CU), and this check makes sense to me (it's only worth trying to combine CUs if there is more than one CU), so it's probably just the comment that is wrong.